### PR TITLE
Pruned Transducer Stateless2 for Aidatatang-200zh (char-based)

### PR DESCRIPTION
**Pull request by luomingshuang**
I build a pruned transducer stateless2 recipe for aidatatang-200zh. I have check the codes with 1 GPU. And it runs normally. I will find free GPUs to finish it. The codes and the results will be updated on real time. BTW, I do a small change https://github.com/lhotse-speech/lhotse/pull/703 on Lhotse to generate the manifest for this dataset.

### CODE REVIEW

**Comment by danpovey** [comment]
cool, do we know what baseline numbers for this dataset are?

**Comment by luomingshuang** [comment]
> cool, do we know what baseline numbers for this dataset are?
Yes, add.

**Comment by luomingshuang** [comment]
BTW, according to the tensorboard log https://tensorboard.dev/experiment/xS7kgYf2RwyDpQAOdS8rAA/#scalars, it seems the loss has not converged completely after training for 30 epochs.
c7ce24c3bd3872f34ae209185e4eea0

**Comment by csukuangfj** [comment]
From the posted results, the total number of words between espnet and icefall is not the same for dev and test.
Is there a reason?

**Comment by luomingshuang** [comment]
This is based on icefall:
`
2022-05-14 21:38:27,533 INFO [utils.py:406] [DEV-beam_size_4] %WER 5.28% [24788 / 469048, 2520 ins, 2576 del, 19692 sub ]
`
`
2022-05-14 22:25:28,530 INFO [utils.py:406] [TEST-beam_size_4] %WER 6.32% [59292 / 937866, 5632 ins, 5648 del, 48012 sub ]
`

**Comment by luomingshuang** [comment]
I doubt that I add the blank to it when compute wer.

**Comment by csukuangfj** [comment]
Ok, I looked at your decoding directory:

For fast beam search:
```
2022-05-15 18:26:53,579 INFO [utils.py:406] [DEV-beam_4_max_contexts_4_max_states_8] %WER 5.29% [24790 / 469048, 2128 ins, 3032 del, 19630 sub ]

2022-05-15 19:25:32,882 INFO [utils.py:406] [TEST-beam_4_max_contexts_4_max_states_8] %WER 6.33% [59408 / 937866, 4674 ins, 6946 del, 47788 sub ]
```

Both espent and kaldi have 468933 words for test and espnet has 234524 for dev.

But in your results, there are 937866 words for test and 469048 words for dev.

**Comment by csukuangfj** [comment]
Your total number of words is doubled.

468933 * 2 = 937866

234524 * 2 = 469048

**Comment by csukuangfj** [comment]
> I doubt that I add the blank to it when compute wer.

Where is the code for this?

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+
+    if params.decoding_method == "greedy_search":
+        log_interval = 100
+    else:
+        log_interval = 50
+
+    results = defaultdict(list)
+    for batch_idx, batch in enumerate(dl):
+        texts = batch["supervisions"]["text"]
+        texts = [list(str(text).replace(" ", "")) for text in texts]
```
[the remark this answers is not in the document]
Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:369]: One of the original texts shows as `晚上 准备 怎么 过 呢`. It has the blank for splitting words.

**Comment by csukuangfj** [comment]
```
2022-05-15 19:23:52,609 INFO [decode.py:395] batch 1148/?, cuts processed until now is 94833
2022-05-15 19:24:00,041 INFO [decode.py:395] batch 1150/?, cuts processed until now is 94939
2022-05-15 19:24:07,416 INFO [decode.py:395] batch 1152/?, cuts processed until now is 95045
2022-05-15 19:24:14,814 INFO [decode.py:395] batch 1154/?, cuts processed until now is 95152
2022-05-15 19:24:22,184 INFO [decode.py:395] batch 1156/?, cuts processed until now is 95260
2022-05-15 19:24:29,516 INFO [decode.py:395] batch 1158/?, cuts processed until now is 95368
2022-05-15 19:24:36,821 INFO [decode.py:395] batch 1160/?, cuts processed until now is 95476
2022-05-15 19:24:44,138 INFO [decode.py:395] batch 1162/?, cuts processed until now is 95584
2022-05-15 19:24:51,383 INFO [decode.py:395] batch 1164/?, cuts processed until now is 95693
2022-05-15 19:24:58,637 INFO [decode.py:395] batch 1166/?, cuts processed until now is 95803
2022-05-15 19:25:05,851 INFO [decode.py:395] batch 1168/?, cuts processed until now is 95913
2022-05-15 19:25:13,059 INFO [decode.py:395] batch 1170/?, cuts processed until now is 96023
2022-05-15 19:25:20,253 INFO [decode.py:395] batch 1172/?, cuts processed until now is 96133
2022-05-15 19:25:27,399 INFO [decode.py:395] batch 1174/?, cuts processed until now is 96238
2022-05-15 19:25:31,775 INFO [decode.py:412] The transcripts are stored in pruned_transducer_stateless2/exp/fast_beam_search/recogs-TEST-beam_4_max_contexts_4_max_states_8-epoch-29-avg-19-beam-4-max-contexts-4-max-states-8.txt
2022-05-15 19:25:32,882 INFO [utils.py:406] [TEST-beam_4_max_contexts_4_max_states_8] %WER 6.33% [59408 / 937866, 4674 ins, 6946 del, 47788 sub ]
2022-05-15 19:25:35,727 INFO [decode.py:425] Wrote detailed error stats to pruned_transducer_stateless2/exp/fast_beam_search/errs-TEST-beam_4_max_contexts_4_max_states_8-epoch-29-avg-19-beam-4-max-contexts-4-max-states-8.txt
2022-05-15 19:25:35,734 INFO [decode.py:442]
For TEST, WER of different settings are:
beam_4_max_contexts_4_max_states_8	6.33	best for TEST

2022-05-15 19:25:35,734 INFO [decode.py:599] Done!
```

It is not the issue of blanks. You have decoded twice as much data as the test and dev dataset.
You can see that the number of processed cuts is doubled from the log.

Please recheck your code for loading test and dev data.

**Comment by luomingshuang** [comment]
Oh, I will have a check.

**Comment by csukuangfj** [comment]
From https://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/iterable_dataset.py#L19
```
    .. caution: With an iterable-style dataset, the sampler replicas exist in the
        dataloading worker subprocesses. That means unless you take extra steps to avoid
        data duplication, it may happen that each worker returns exactly the same data.
        This problem is avoided with WebDataset by using sharding -- we let WebDataset
        subset the shards visible in each subprocess (and each node in multi-GPU DDP training).
```

Please change number of workers from 2 to 1 in your case. That is, use
```
--num-workers=1
```
or
```
--num-workers=0
```
during decoding.

The default value is 2. That is why your number of cuts is doubled.

**Comment by pzelasko** [comment]
Your setup doesn't seem to use webdataset, in that case I don't think it makes too much sense to use `IterableDatasetWrapper`, just pass the dataset and sampler straight to DataLoader (and you can remove manual settings on rank and world size in the sampler).

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+    num_param = sum([p.numel() for p in model.parameters()])
+    logging.info(f"Number of model parameters: {num_param}")
+
+    # Note: Please use "pip install webdataset==0.1.103"
+    # for installing the webdataset.
+    import glob
+    import os
+
+    from lhotse import CutSet
+    from lhotse.dataset.webdataset import export_to_webdataset
```
Comment (luomingshuang) [inline egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: Oh, I use webdataset here.

Comment (pzelasko) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: Hmm interesting. Were you running multi-GPU decoding? In that case you might find that updating to latest Lhotse will help, I recently fixed an issue that only appears with multi-GPU and multi-dataloader-worker scripts.

Comment (csukuangfj) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: > Were you running multi-GPU decoding?

I don't think so. All the decoding code in icefall uses a single GPU.

Is it expected that if `--num-workers=2` in this case, the data is doubled?

Comment (pzelasko) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: No, that shouldn't happen since he used the shards option in exporting to webdataset. If you can confirm it's really because of webdataset, and updating to latest Lhotse (and webdataset) doesn't help, we'll debug it.

Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: OK.

Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: I use the latest Lhotse and webdataset to decode again. When I set num-workers as 2 and a single GPU, the data is still doubled.

Comment (csukuangfj) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: Also, it works fine when number of workers is 1 or 0, I think.

Comment (pzelasko) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: OK I think what happened. Replace `BucketingSampler` with `DynamicBucketingSampler` and it should resolve the issues. With `BucketingSampler`, the cuts are eagerly read into memory before they go into dataloader's sub-processes, so webdataset doesn't know to de-duplicate them. I believe you are seeing a warning about BucketingSampler not being a good fit for lazy cutsets; now I'm thinking I should make it raise an exception saying to migrate to DynamicBucketingSampler in these cases to avoid such issues.

Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: Ah, yes. `luomingshuang/anaconda3/envs/k2-python/lib/python3.8/site-packages/lhotse-1.2.0.dev0+git.de75634.clean-py3.8.egg/lhotse/dataset/sampling/bucketing.py:102: UserWarning: Lazy CutSet detected in BucketingSampler: we will read it into memory anyway. Please use lhotse.dataset.DynamicBucketingSampler instead.
`

Comment (pzelasko) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: See: https://github.com/lhotse-speech/lhotse/pull/710

Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: But, isn't DynamicBucketSampler slow down when loading data?

Comment (pzelasko) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: I think you are referring to issue https://github.com/lhotse-speech/lhotse/issues/678 -- here it's a different use-case, because the sampler itself is moved to the dataloading worker process via `IterableDatasetWrapper`, so any overhead introduced by it doesn't affect the main training loop.

Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py:536]: OK, when using DynamicbucketSampler for decoding, the number of cuts is normal (not doubled) even if the num-workers is 2.
